official/resnet/ctl/ctl_imagenet_main.py

```diff
@@ -0,0 +1,295 @@
+      datasets_num_private_threads=flags_obj.datasets_num_private_threads,
+      dtype=dtype)
+
+  test_ds = None
+  if not flags_obj.skip_eval:
+    test_ds = input_fn(
+        is_training=False,
+        data_dir=flags_obj.data_dir,
+        batch_size=flags_obj.batch_size,
+        num_epochs=flags_obj.train_epochs,
+        parse_record_fn=parse_record_keras,
+        dtype=dtype)
+
+    if strategy:
+      test_ds = strategy.experimental_distribute_dataset(test_ds)
+      steps_per_eval = IMAGENET_VALIDATION_IMAGES // flags_obj.batch_size
+
+  if strategy:
+    train_ds = strategy.experimental_distribute_dataset(train_ds)
+  steps_per_epoch = APPROX_IMAGENET_TRAINING_IMAGES // flags_obj.batch_size
+  train_epochs = flags_obj.train_epochs
+
+  if flags_obj.train_steps:
+    train_steps = min(flags_obj.train_steps, steps_per_epoch)
+    train_epochs = 1
+
+  strategy_scope = distribution_utils.get_strategy_scope(strategy)
+  with strategy_scope:
+    logging.info('Building Keras ResNet-50 model')
+    model = resnet_model.resnet50(num_classes=imagenet_main.NUM_CLASSES,
+                                  dtype=dtype, batch_size=flags_obj.batch_size)
+
+    optimizer = tf.keras.optimizers.SGD(
+        learning_rate=_BASE_LEARNING_RATE, momentum=0.9, nesterov=True)
+
+    training_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
+        'training_accuracy', dtype=tf.float32)
+    test_loss = tf.keras.metrics.Mean('test_loss', dtype=tf.float32)
+    test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
+        'test_accuracy', dtype=tf.float32)
+    logging.info('Finished building Keras ResNet-50 model')
+
+    def train_step(train_ds_inputs):
+      """Training StepFn."""
+      def step_fn(inputs):
+        """Per-Replica StepFn."""
+        images, labels = inputs
+        with tf.GradientTape() as tape:
+          logits = model(images, training=True)
+
+          prediction_loss = tf.keras.losses.sparse_categorical_crossentropy(
+              labels, logits)
+          loss1 = tf.reduce_sum(prediction_loss) * (1.0/ flags_obj.batch_size)
+          loss2 = tf.reduce_sum(model.losses) / strategy.num_replicas_in_sync
+          loss = loss1 + loss2
+
+        grads = tape.gradient(loss, model.trainable_variables)
+        optimizer.apply_gradients(zip(grads, model.trainable_variables))
+
+        training_accuracy.update_state(labels, logits)
+        return loss
+
+      if strategy:
+        per_replica_losses = strategy.experimental_run_v2(
+            step_fn, args=(train_ds_inputs,))
+        return strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses,
+                               axis=None)
+      else:
+        return step_fn(train_ds_inputs)
+
+    def test_step(test_ds_inputs):
+      """Evaluation StepFn."""
+      def step_fn(inputs):
+        images, labels = inputs
+        logits = model(images, training=False)
+        loss = tf.keras.losses.sparse_categorical_crossentropy(labels,
```

Comment:
i remember there was some discussion around which loss to report (both for training and eval): the one scaled by number of replicas or not. What does keras fit do? 
cc @pavithrasv

Reply:
This reports a loss value that is equal to what the user would see if they did not use distribution strategy. I followed the same process that we suggested to our users in the tutorials. 
@pavithrasv can maybe talk more about loss reporting in keras.

Reply:
fit reports the value scaled by number of replicas. Do we want to change that in Keras callbacks?

Reply:
Does this mean that if a user uses DistributionStrategy they will see different loss values? will that be confusing to the user? 
Is it a big change to report the prediction loss instead of the scaled one?

Reply:
ideally, we report the same loss in fit and ctl and with and without strategy.
for this PR, it's fine but let's try to come to some consensus on this topic as users do get confused when they say different loss being reported.